### PR TITLE
WIP: 128 hash table mock

### LINKED ISSUE #35: use a collection that allows more efficient inserting of pairs into source chains
currently we insert new pairs at the start of a vector, which is O(n)

other collection types can do O(1) or O(log n) for inserts

https://doc.rust-lang.org/std/collections/#sequences

### LINKED ISSUE #128: Mock local hash table
child of #88 

parent issue is about mocking the whole DHT, this is just the HT part

### PR DESCRIPTION
**Pull request by thedavidmeister**
fixes #128 
fixes #35 

followup:

- thread safe agent state #137 
- better error handling in Header::new #147
- txn ordering for PairMeta #138 
- signing for PairMeta #139 
- handle incoming PairMeta #140 
- thread safe tables #135 
- error handling in chain iterator #146
- handle failed meta assertions #142 
- review sorting of return value from pair meta #144 
- queries, in general #141 
- review the meta attribute names for crud statuses #143 
- better handling for generic types in state #148

### CODE REVIEW

File: core/src/chain/mod.rs [hunk 1]
```diff
@@ -1,31 +1,222 @@
...
+    table: Rc<HashTable>,
+    current: Option<Pair>,
 
-    /// returns an iterator referencing pairs from top (most recent) to bottom (genesis)
-    fn iter(&self) -> std::slice::Iter<Pair>;
+}
+
+impl ChainIterator {
+
+    pub fn new<HT: 'static + HashTable> (table: Rc<HashTable>, pair: &Option<Pair>) -> ChainIterator {
```
Comment (sphinxc0re) [inline core/src/chain/mod.rs]: I think the Iterator should do the same thing as the Chain with the HashTable type.

File: core/src/chain/mod.rs [hunk 2]
```diff
@@ -1,31 +1,222 @@
...
+
+    pub fn table(&self) -> Rc<T> {
+        Rc::clone(&self.table)
+    }
+
+    pub fn push (&mut self, entry: &Entry) -> Result<Pair, HolochainError> {
+        let pair = Pair::new(self, entry);
+
+        if !(pair.validate()) {
+            return Result::Err(HolochainError::new("attempted to push an invalid pair for this chain"))
```
Comment (sphinxc0re) [inline core/src/chain/mod.rs]: strip the `Result::`

File: core/src/chain/mod.rs [hunk 3]
```diff
@@ -1,31 +1,222 @@
...
+        }
+
+        let top_pair = self.top().and_then(|p| Some(p.key()));
+        let next_pair = pair.header().next();
+
+        if top_pair != next_pair {
+            return Result::Err(HolochainError::new(
```
Comment (sphinxc0re) [inline core/src/chain/mod.rs]: strip the `Result::`

File: core/src/chain/mod.rs [hunk 4]
```diff
@@ -1,31 +1,222 @@
...
+        // @TODO implement incubator for thread safety
+        // @see https://github.com/holochain/holochain-rust/issues/135
+        let table = Rc::get_mut(&mut self.table).unwrap();
+        let result = table.commit(&pair);
+        if result.is_ok() {
+            self.top = Some(pair.clone());
+        }
+        match result {
+            Result::Ok(_) => Result::Ok(pair),
+            Result::Err(e) => Result::Err(e),
```
Comment (sphinxc0re) [inline core/src/chain/mod.rs]: Same here: strip the `Result::`

File: core/src/hash_table/memory/mod.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    /// smoke test
+    fn new() {
+        test_table();
+    }
+
+    #[test]
+    /// tests for ht.open()
+    fn open() {
+        let mut ht = test_table();
+        assert_eq!(Result::Ok(()), ht.open());
```
Comment (sphinxc0re) [inline core/src/hash_table/memory/mod.rs]: strip the `Result::`

File: core/src/hash_table/memory/mod.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    #[test]
+    /// tests for ht.close()
+    fn close() {
+        let mut ht = test_table();
+        assert_eq!(Result::Ok(()), ht.close());
```
Comment (sphinxc0re) [inline core/src/hash_table/memory/mod.rs]: strip the `Result::`

File: core/src/hash_table/pair_meta.rs [hunk 7]
```diff
@@ -154,4 +157,31 @@ pub mod tests {
...
+        assert_eq!(Ordering::Greater, m_2ax.cmp(&m_1ax));
+        assert_eq!(Ordering::Less, m_1ay.cmp(&m_2ax));
+
+        // sort by attribute key
+        assert_eq!(Ordering::Less, m_1ax.cmp(&m_1bx));
+        assert_eq!(Ordering::Greater, m_1bx.cmp(&m_1ax));
+
+        // sort by attribute value
+        assert_eq!(Ordering::Less, m_1ax.cmp(&m_1ay));
+        assert_eq!(Ordering::Greater, m_1ay.cmp(&m_1ax));
```
Comment (sphinxc0re) [inline core/src/hash_table/pair_meta.rs]: Wouldn't it be better to assert for the actual condition? So e.g.
```rust
assert!(m_1ax < m_2ax);
assert_eq!(m_1ax, m_1ax);
assert!(m_2ax > m_1ax);
assert!(m_1ay < m_2ax);
```

Comment (thedavidmeister) [reply core/src/hash_table/pair_meta.rs]: well asserting equality is testing a different thing, as for the others i'm not sure

it makes sense to me that i'm testing `cmp` but i could add the operators as well to cover both

**Pull request merged**